acl-sandbox/group_theory/fixing_subgroup.lean

```diff
@@ -0,0 +1,293 @@
+/-
+Copyright (c) 2022 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import tactic
+
+import group_theory.subgroup.basic
+import group_theory.group_action.basic
+import order.order_dual
+
+/-!
+
+# Fixing submonoid, fixing subgroup of an action
+
+In the presence of of an action of a monoid or a group,
+this file defines the fixing submonoid or the fixing subgroup,
+and relates it to the set of fixed points via a Galois connection.
+
+## Main definitions
+
+* `fixing_submonoid M s` :
+in the presence of `mul_action M α` (with `monoid M`)
+it is the `submonoid M` consisting of elements which fix `s : set α` pointwise.
+
+* `fixing_submonoid_fixed_points_connection M α` is the `galois_connection`
+that relates `fixing_submonoid` with `fixed_points`.
+
+* `fixing_submonoid_of_union` and `fixing_submonoid_of_Union` are consequences
+of the Galois connection, as well as `fixed_points_of_sup` and `fixed_points_of_supr`.
+
+* `fixing_subgroup M s` :
+in the presence of `mul_action M α` (with `group M`)
+it is the `subgroup M` consisting of elements which fix `s : set α` pointwise.
+
+* `fixing_subgroup_fixed_points_connection M α` is the `galois_connection`
+that relates `fixing_subgroup` with `fixed_points`.
+
+* `fixing_subgroup_of_union` and `fixing_subgroup_of_Union` are consequences
+of the Galois connection,
+as well as `fixed_points_of_group_of_sup` and `fixed_points_of_group_of_supr`.
+
+* The file starts with some lemmas that allow to rewrite `antitone` into `monotone`
+in various cases.
+Apparently the `monotone.dual` (and analogues) were not sufficient, so I wrote `monotone.dual_iff`
+that works in both directions.
+
+TODO :
+
+* Adjust names
+
+* Remove the definitions with the same names from the Galois theory file
+
+* Decide what needs to be done with the antitonicity section. In particular, the
+proofs need two directions, but the proof term is exactly the same (up to types).
+
+* Maybe other lemmas are useful
+
+* Treat semigroups ?
+
+-/
+
+section antitonicity
+
+universes u v
+variables {α : Type u} {β : Type v} [preorder α] [preorder β] {f : α → β}
+
+open order_dual
+
+lemma monotone.dual_iff : monotone f ↔  monotone (to_dual ∘ f ∘ of_dual) :=
+begin
+  split; exact λ hf a b h, hf h
+end
+
+lemma monotone.dual_left_iff : monotone f ↔ antitone (f ∘ of_dual) :=
+begin
+  split; exact λ hf a b h, hf h
+end
+
+lemma monotone.dual_right_iff : monotone f ↔ antitone (to_dual ∘ f) :=
+begin
+  split; exact λ hf a b h, hf h
+end
```

Comment:
```suggestion
@[simp] lemma antitone_to_dual_comp : antitone (to_dual ∘ f) ↔ monotone f := iff.rfl
```
The rule is that postcomposing (on the right) makes it `iff.rfl` while precomposing (on the left) makes it `forall_swap`.

Reply:
huh!

Reply:
@YaelDillies : 
There was an issue that the file in which you proposed the commit does not exist anymore. 
Do you mean all previous lemmas have to be deleted and only your two simp-lemmas suffice?

Reply:
Also : the name of the first lemma should end in _iff…

Reply:
The forall2-swap lemma I put in logic.basic in a parallel PR, or just now in that one ? (Isn't it simpler?)

Reply:
Ah whoops. You should have four simp lemmas in the end. You should PR these four simp lemmas (+ the `monotone_on`/`antitone_on` versions + `forall₂_swap`) in a separate PR.

Reply:
(The forall2-swap lemma does not work… bizarrely, I need something else.)

Reply:
What works is 
```
@[simp] lemma monotone_on_comp_of_dual_iff : monotone_on (f ∘ of_dual) s ↔ antitone_on f s := 
-- forall₂_swap
begin
split;
exact λ hf a ha b hb h, hf hb ha h
end 
```

Reply:
See #13207 for the PR that gives these simp lemmas.